builtin/providers/google/resource_storage_bucket.go

```diff
@@ -0,0 +1,149 @@
+
+	// Get the bucket and acl
+	bucket := d.Get("bucket").(string)
+	acl := d.Get("predefined-acl").(string)
+	location := d.Get("location").(string)
+
+	// Create a bucket, setting the acl, location and name.
+
+	sb := &storage.Bucket{Name: bucket, Location: location}
+	res, err := config.clientStorage.Buckets.Insert(config.Project, sb).PredefinedAcl(acl).Do()
+
+	if err == nil {
+		log.Printf("[DEBUG] Created bucket %v at location %v\n\n", res.Name, res.SelfLink)
+	} else {
+		fmt.Printf("Error creating bucket %s: %v", bucket, err)
+	}
+
+	// Assign the bucket name as the resource ID
+	// Add extra metadata about the bucket.
+	d.SetId(res.Id)
+	// Hang on do we want to do this here?
+	// We want to set the ACL based on the schema
+	// definition - therefore if we read the
+	// ACL at create point, we risk never
+	// setting the intended ACL.
+	d.Set("predefined-acl", acl)
+	d.Set("location", location)
+
+	return nil
+}
+
+func resourceStorageBucketUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	if d.HasChange("predefined-acl") {
+		fmt.Printf("Error you can't change the ACL of a bucket currently.")
+	}
+
+	if d.HasChange("location") {
+		fmt.Printf("Error you can't change the location of a bucket in GCS.")
+	}
+
+	return nil
+}
+
+func resourceStorageBucketRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// Get the bucket and acl
+	bucket := d.Get("bucket").(string)
+	res, err := config.clientStorage.Buckets.Get(bucket).Do()
+	if err == nil {
+		log.Printf("[DEBUG] Read bucket %v at location %v\n\n", res.Name, res.SelfLink)
+	} else {
+		fmt.Printf("Error reading bucket %s: %v", bucket, err)
+	}
+
+	// Assign the bucket name as the resource ID
+	// Add extra metadata about the bucket.
+	d.SetId(res.Id)
+	d.Set("predefined-acl", res.Acl)
+	d.Set("location", res.Location)
+	return nil
+}
+
+func resourceStorageBucketDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// Get the bucket
+	bucket := d.Get("bucket").(string)
+	if res, err := config.clientStorage.Objects.List(bucket).Do(); err == nil {
+		if err != nil {
+			fmt.Printf("Error Objects.List failed: %v", err)
+		}
+		if len(res.Items) != 0 {
+			if d.Get("force_destroy").(bool) {
+				// bucket may have things delete them if forceDestroy flag is set.
+				log.Printf("[DEBUG] GCS Bucket attempting to forceDestroy\n\n")
+				for _, object := range res.Items {
+					log.Printf("[DEBUG] Found %s", object.Name)
+					if err := config.clientStorage.Objects.Delete(bucket, object.Name).Do(); err != nil {
+						log.Fatalf("Error trying to delete object: %s %s\n\n", object.Name, err)
+					} else {
+						log.Printf("Object deleted: %s \n\n", object.Name)
+
+					}
+				}
+			} else {
+				delete_err := errors.New("Error trying to delete a bucket containing objects without `force_destroy` set to true")
+				log.Printf("Error! %s : %s\n\n", bucket, delete_err)
+				return delete_err
+			}
+		}
```

Comment:
I think there's a need for a loop to delete all objects in the bucket in case there's more than 1000 (which is maximum a single google storage api call will return) and then as an improvement, probably batch all object deletes together (I think you can batch up to 1000 operations, so this is within the limit).